### PR TITLE
Check equality constraints when checking whether a constraint is satisfied

### PR DESCRIPTION
**Pull request by zygoloid**
Add checking that a type only satisfies a constraint if it satisfies all of that constraint's equality and rewrite constraints.

Enforce the rule that `=` must be used in impls when specifying associated constant values rather than `==`.

Turn off the pre-#2173 single-step equality behavior. This is getting somewhat ahead of the approved design, but it's a one-line change to restore the old behavior.

Fix `CARBON_CHECK` to handle top-level `,`s in its argument, such as may happen in template argument lists, as this change introduces such a check.

### CODE REVIEW

File: explorer/interpreter/impl_scope.cpp [hunk 1]
```diff
@@ -103,7 +103,33 @@ auto ImplScope::Resolve(Nonnull<const Value*> constraint_type,
...
+      for (auto& equal : equals) {
+        auto it = equal.values.begin();
+        Nonnull<const Value*> first =
+            type_checker.Substitute(local_bindings, *it++);
+        for (; it != equal.values.end(); ++it) {
+          Nonnull<const Value*> current =
+              type_checker.Substitute(local_bindings, *it);
+          if (!ValueEqual(first, current, &equality_ctx)) {
+            return ProgramError(source_loc) << "could not determine that "
+                                            << *first << " == " << *current;
```
Comment (jonmeow) [inline explorer/interpreter/impl_scope.cpp]: Looking at `could not determine that 4 == 5` I'm wondering if more words may help explain what's going on. Like, can we instead say something like "Requires that constraints 4 and 5 are equal, which isn't provable."?

File: explorer/interpreter/impl_scope.cpp [hunk 2]
```diff
@@ -233,4 +259,49 @@ void ImplScope::Print(llvm::raw_ostream& out) const {
...
+                   << *impl_scope_;
+  }
+
+  if (!impl_scope_->VisitEqualValues(value, visitor)) {
+    return false;
+  }
+
+  // Also look up and visit the corresponding impl if this is an associated
+  // constant.
+  // TODO: Is this necessary?
```
Comment (jonmeow) [inline explorer/interpreter/impl_scope.cpp]: In what way aren't you sure? Could this be a CHECK assuming it's not necessary, maybe leaving code commented out if it turns out to be?

File: common/check.h [hunk 3]
```diff
@@ -21,18 +21,18 @@ namespace Carbon {
...
 // For example:
 //   CARBON_CHECK(is_valid) << "Data is not valid!";
-#define CARBON_CHECK(condition)                                           \
-  (condition) ? (void)0                                                   \
-              : CARBON_RAW_EXITING_STREAM()                               \
-                    << "CHECK failure at " << __FILE__ << ":" << __LINE__ \
-                    << ": " #condition                                    \
-                    << Carbon::Internal::ExitingStream::AddSeparator()
+#define CARBON_CHECK(...)                                                   \
+  (__VA_ARGS__) ? (void)0                                                   \
```
Comment (jonmeow) [inline common/check.h:25]: Can you please explain this change in your PR description? While I'm fine with it, it seems unrelated to what the PR description currently explains and significant enough that it should be documented.

File: explorer/interpreter/interpreter.cpp [hunk 4]
```diff
@@ -715,9 +735,12 @@ auto Interpreter::Convert(Nonnull<const Value*> value,
               InstantiateType(destination_type, source_loc));
           return arena_->New<NominalClassValue>(inst_dest, value);
         }
-        default:
-          CARBON_FATAL() << "Can't convert value " << *value << " to type "
-                         << *destination_type;
+        default: {
+          CARBON_CHECK(isa<VariableType, AssociatedConstant>(destination_type))
```
Comment (jonmeow) [inline explorer/interpreter/interpreter.cpp]: You're doing this exact same multi-check in multiple places. Elsewhere, in the AST, we use parent classes to handle situations where multiple features behave identically. But I think it's actually particularly odd here: this is already switching on the Kind. If you need these to behave differently, not add a case?

Also, did this need to start handling VariableType as some consequence of your changes? I understand AssociatedConstant here because it's the focus of your changes, but VariableType seems a bit of an unusual side-effect and the reason why isn't obvious to me.

File: explorer/interpreter/interpreter.cpp [hunk 5]
```diff
@@ -747,9 +770,12 @@ auto Interpreter::Convert(Nonnull<const Value*> value,
                                            &array_type.element_type());
           break;
         }
-        default:
-          CARBON_FATAL() << "Can't convert value " << *value << " to type "
-                         << *destination_type;
+        default: {
+          CARBON_CHECK(isa<VariableType, AssociatedConstant>(destination_type))
```
Comment (jonmeow) [inline explorer/interpreter/interpreter.cpp]: As above: when you're already switching on the Kind, shouldn't we prefer cases?

File: explorer/interpreter/type_checker.h [hunk 6]
```diff
@@ -79,8 +79,12 @@ class TypeChecker {
                                    int impl_offset) const
       -> Nonnull<const Witness*>;
 
+  // Returns the trace stream, if one was provided.
+  auto trace_stream() const -> std::optional<Nonnull<llvm::raw_ostream*>> {
```
Comment (jonmeow) [inline explorer/interpreter/type_checker.h]: This feels a bit odd to add as a public API: I don't really think that a trace stream fits with type checking APIs... Of course, now I'm looking at the other things made public here, and maybe this is no longer really a basic TypeChecker API, but has morphed into some mix of public type checking API and internal utility holder?

File: common/check.h (hunk 3 above)
Comment (zygoloid) [reply common/check.h:25]: Done.

File: explorer/interpreter/impl_scope.cpp (hunk 2 above)
Comment (zygoloid) [reply explorer/interpreter/impl_scope.cpp]: Sorry, that was supposed to be a note to myself to go back and check!

Having checked, this is indeed unnecessary. Removed.

File: explorer/interpreter/interpreter.cpp (hunk 4 above)
Comment (zygoloid) [reply explorer/interpreter/interpreter.cpp]: Factored out the duplicated check into a named function.

This isn't a consequence of my change; it's just that no-one had written a test that encountered this situation before.

File: explorer/interpreter/interpreter.cpp (hunk 5 above)
Comment (zygoloid) [reply explorer/interpreter/interpreter.cpp]: Let me know if you're unhappy with the new approach.

File: explorer/interpreter/type_checker.h (hunk 6 above)
Comment (zygoloid) [reply explorer/interpreter/type_checker.h]: Yeah, I wonder if we should really have a single monolithic `TypeChecker` class at all. I think a different design for passing around the arena and trace stream would likely be better, and should allow us to make a lot of things that are currently `TypeChecker` members be independent functions / classes.

Short-term, I agree that this doesn't make a lot of sense as public API, and having removed the code with the TODO from `SingleStepEqualityContext` I'm not sure we really need this at all any more, so... removed.

File: explorer/interpreter/impl_scope.cpp (hunk 1 above)
Comment (zygoloid) [reply explorer/interpreter/impl_scope.cpp]: Wording improved. I'm avoiding saying "which isn't provable" because there are cases where such things are provable but we nonetheless can't discover the proof for ourselves. I've also added a test with an example of that.

File: explorer/interpreter/value.h [hunk 7]
```diff
@@ -121,6 +121,13 @@ class Value {
   const Kind kind_;
 };
 
+// Returns whether the fully-resolved kind that this value will eventually have
+// is currently unknown, because it depends on a generic parameter.
+inline bool IsValueKindDependent(Nonnull<const Value*> type) {
```
Comment (jonmeow) [inline explorer/interpreter/value.h:126]: FWIW I think this name a bit hard to interpret, maybe "IsDependentType" similar to "IsConcreteType" or "IsType"?

**Review by jonmeow — APPROVED** [review]

File: explorer/interpreter/value.h (hunk 7 above)
Comment (zygoloid) [reply explorer/interpreter/value.h:126]: Hm, this doesn't tell you whether the type is dependent, at least in the C++ sense. For example, `Vector(T)` would probably be considered a dependent type, but the value kind (`NominalClassType`) is not dependent -- this is specifically determining whether the `Value::Kind` is dependent. I'm having a hard time finding a better name that's still appropriate.
